### PR TITLE
Adding Heterogeneous Clusters example for TensorFlow and PyTorch

### PR DESCRIPTION
**Pull request by gilinachum**
*Description of changes:* Heterogeneous Clusters for Amazon SageMaker model training was [announced July 2022](https://aws.amazon.com/about-aws/whats-new/2022/07/announcing-heterogeneous-clusters-amazon-sagemaker-model-training/). We're providing extensive code examples for using the feature with TensorFlow and PyTorch.

*Testing done:* Run the TensorFlow and PyTorch notebooks (2 in total).

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

- [x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-sagemaker-examples/blob/master/CONTRIBUTING.md) doc and adhered to the example notebook best practices
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-sagemaker-examples/blob/master/README.md)
- [x] I have tested my notebook(s) and ensured it runs end-to-end
- [x] I have linted my notebook(s) and code using `black-nb -l 100 {path}/{notebook-name}.ipynb`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: training/heterogeneous-clusters/tf.data.service.sagemaker/code/launcher.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    print(f'Opening process async: {params}')
+    p = subprocess.Popen(params)
+    print(f'Process {name} started')
+    return p.pid
+
+
+def start_child_process(name : str, additional_args=[]) -> int:
+     params = ["python", f"./{name}"] + sys.argv[1:] + additional_args
+     print(f'Opening process: {params}')
+     p = subprocess.run(params)
```
[the remark this answers is not in the document]
Comment (gilinachum) [reply training/heterogeneous-clusters/tf.data.service.sagemaker/code/launcher.py:24]: This warning is not relevant in the current context which is executing the original command line SageMaker training toolkit ask for, just with an additional parameter.  
How do I add a relevant ignore?

File: sagemaker-datawrangler/readme.md [hunk 2]
```diff
@@ -1 +1,41 @@
+![Amazon SageMaker Data Wrangler](https://github.com/aws/amazon-sagemaker-examples/raw/main/_static/sagemaker-banner.png)
```
Comment (aaronmarkham) [inline sagemaker-datawrangler/readme.md:1]: Not sure why this change is here... seems to be the same file as what's in the repo now.

File: training/heterogeneous-clusters/hello.world.sagemaker/README.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
+# Hetero Training Job - Hello world
```
Comment (aaronmarkham) [inline training/heterogeneous-clusters/hello.world.sagemaker/README.md:1]: Please add an index.rst and link it up to the training/index.rst page, so your content shows up on the website.
More info is here: https://quip-amazon.com/7NdEATw7Hlbb/Adding-new-large-example-projects-to-the-Examples-website

File: training/heterogeneous-clusters/hello.world.sagemaker/start_job.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import datetime
+from sagemaker.tensorflow import TensorFlow
+from sagemaker.instance_group import InstanceGroup
+import os
+
+REGION = 'us-east-1'
```
Comment (aaronmarkham) [inline training/heterogeneous-clusters/hello.world.sagemaker/start_job.py:6]: We've recently started doing multi-region testing, so I imagine this is going to break. Does this solution only work in us-east-1?

File: training/heterogeneous-clusters/hello.world.sagemaker/start_job.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+os.environ["AWS_DEFAULT_REGION"] = REGION
+
+# https://aws.amazon.com/sagemaker/pricing/
+data_group = InstanceGroup("data_group", "ml.c5.xlarge", 1)
```
Comment (aaronmarkham) [inline training/heterogeneous-clusters/hello.world.sagemaker/start_job.py:10]: Hard-coding the instance types may also break across certain regions.

File: training/heterogeneous-clusters/pt.grpc.local/README.md [hunk 6]
```diff
@@ -0,0 +1,106 @@
+### Local Training using gRPC Client-Server
+
+The example here provides a conceptual idea of how to separate data preprocessing and training processes, and establish a gRPC client server communication between these components. We re-use the same client-server communication implementation in our Heterocluster SageMaker training example <PUT THE LINK when we move to aws-samples>. To demonstrate this we first run baseline testing where both the components (data preprocessing and training) run in a set of processes (see section A). In Section B, we split data processing run on gRPC Server set of processes whereas training run on a different set of processes.
```
Comment (aaronmarkham) [inline training/heterogeneous-clusters/pt.grpc.local/README.md:3]: You have a temp link here.

File: training/heterogeneous-clusters/pt.grpc.local/README.md [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+Important:
```
Comment (aaronmarkham) [inline training/heterogeneous-clusters/pt.grpc.local/README.md:5]: You should probably also have this information on an index.rst or in a notebook. As far as I can tell now, it's not testable, which can be an issue over time. How do you assure quality of this example? Can you put these commands and info in a notebook, so the flow can be tested?

File: training/heterogeneous-clusters/pt.grpc.local/README.md [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+```
+python main_grpc_server.py --batch-size 8192 --num-data-workers 4 --iterations 100 --grpc-workers 2
+```
+where,  
+`batch-size` any integer  
+`num-workers` based on no. of cpu per of your data pre-processing instance   
+`iterations` no. of iterations in an epoch - must be multiple of 10s`   
+`grpc-workers` no. of workers fetching the pre-processed data to DNN process (gRPC client)   
+
+**Step 2**: Run gRPC Client in a new terminal session (File > New > Terminal) . The set of process spawend by this script fetches pre-processed data from server and runs training. Make sure you change your working directory to where code exist (..\..\pt.grpc.local).
```
Comment (aaronmarkham) [inline training/heterogeneous-clusters/pt.grpc.local/README.md:64]:
```suggestion
**Step 2**: Run gRPC Client in a new terminal session (File > New > Terminal) . The set of process spawned by this script fetches pre-processed data from server and runs training. Make sure you change your working directory to where code exist (..\..\pt.grpc.local). 
```

File: training/heterogeneous-clusters/pt.grpc.local/README.md [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+`grpc-workers` no. of workers fetching the pre-processed data to DNN process (gRPC client)   
+
+**Step 2**: Run gRPC Client in a new terminal session (File > New > Terminal) . The set of process spawend by this script fetches pre-processed data from server and runs training. Make sure you change your working directory to where code exist (..\..\pt.grpc.local). 
+```
+cd ~/SageMaker/hetro-training/pt.grpc.local/
+python main_grpc_client.py --batch-size 8192 --num-dnn-workers 2 --iterations 100 --model-dir ./
+```
+where,   
+`batch-size` any integer, must match to the size mentioned in the gRPC server process launch   
+`num-workers` no. of dataloader workers, it is based on no. of cpu of the dnn instance
```
Comment (aaronmarkham) [inline training/heterogeneous-clusters/pt.grpc.local/README.md:71]:
```suggestion
`num-workers` no. of dataloader workers, it is based on the number of CPUs of the DNN instance   
```

File: training/heterogeneous-clusters/pt.grpc.local/README.md [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+**Step 2**: Run gRPC Client in a new terminal session (File > New > Terminal) . The set of process spawend by this script fetches pre-processed data from server and runs training. Make sure you change your working directory to where code exist (..\..\pt.grpc.local). 
+```
+cd ~/SageMaker/hetro-training/pt.grpc.local/
+python main_grpc_client.py --batch-size 8192 --num-dnn-workers 2 --iterations 100 --model-dir ./
+```
+where,   
+`batch-size` any integer, must match to the size mentioned in the gRPC server process launch   
+`num-workers` no. of dataloader workers, it is based on no. of cpu of the dnn instance   
+`iterations` no. of iterations in an epoch - must be multiple of 10s, must match to the no. mentioned in the gRPC server process launch
```
Comment (aaronmarkham) [inline training/heterogeneous-clusters/pt.grpc.local/README.md:72]:
```suggestion
`iterations` no. of iterations in an epoch - must be multiple of 10s, and must match the number mentioned in the gRPC server process launch   
```

File: training/heterogeneous-clusters/pt.grpc.local/README.md [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+70: avg step time: 0.6490721581571441
+80: avg step time: 0.7544923407874989
+90: avg step time: 0.8115056776666633
+100: avg step time: 0.8806567812599997
+Saving the model
+Training job completed!
+Shutting down data service via port 16000
+``` 
+    
+**Step 4**: Optionally, in a new terminal window, you can validate whether the gRPC client-server communication is taking place.
```
Comment (aaronmarkham) [inline training/heterogeneous-clusters/pt.grpc.local/README.md:94]: There are security implications throughout (lots of network communications and popping shells), but no security section in the documentation. This should be addressed.

File: training/heterogeneous-clusters/pt.grpc.sagemaker/README.md [hunk 12]
```diff
@@ -0,0 +1,6 @@
+# SageMaker heterogeneous Training ("Hetero") - Pytorch example (MNIST Dataset)
```
Comment (aaronmarkham) [inline training/heterogeneous-clusters/pt.grpc.sagemaker/README.md:1]: Copy content to an index.rst and link to your index.rst page in the upstream folder.

File: training/heterogeneous-clusters/pt.grpc.sagemaker/README.md [hunk 13]
```diff
@@ -0,0 +1,6 @@
...
+This example demonstrates a more general way of offloading pre-processing to auxiliary devices using gRPC, the same protocol underlying the TensorFlow data service. We use here pytorch 1.11 framework. The job is submitted to SageMaker using Hetero feature that allows you to run one training job that includes instances of different types (for example a GPU instance like ml.p3.2xlarge and a CPU instance like c5.9xlarge). The primary use case here is offloading CPU intensive tasks like image pre-processing (data augmentation) from the GPU instance to a dedicate CPU instance, so you can fully utilize the exensive GPU, and arrive at an improved time and cost to train.
```
Comment (aaronmarkham) [inline training/heterogeneous-clusters/pt.grpc.sagemaker/README.md:2]:
```suggestion
This example demonstrates a more general way of offloading pre-processing to auxiliary devices using gRPC, the same protocol underlying the TensorFlow data service. We use here the PyTorch 1.11 framework. The job is submitted to SageMaker using the Heterogenous training feature that allows you to run one training job that includes instances of different types (for example a GPU instance like ml.p3.2xlarge and a CPU instance like c5.9xlarge). The primary use case here is offloading CPU intensive tasks like image pre-processing (data augmentation) from the GPU instance to a dedicate CPU instance, so you can fully utilize the more expensive GPU, and arrive at an improved time and cost to train.
```

File: training/heterogeneous-clusters/pt.grpc.sagemaker/README.md [hunk 14]
```diff
@@ -0,0 +1,6 @@
...
+ 
+
+## Instructions
+Follow steps in [notebook](./hetero-pytorch-mnist.ipynb)
```
Comment (aaronmarkham) [inline training/heterogeneous-clusters/pt.grpc.sagemaker/README.md:6]: For accessibility...
```suggestion
Follow steps in the [heterogenous training with PyTorch notebook](./hetero-pytorch-mnist.ipynb)
```

File: training/heterogeneous-clusters/pt.grpc.sagemaker/code/requirements.txt [hunk 15]
```diff
@@ -0,0 +1,2 @@
+torchvision
```
Comment (aaronmarkham) [inline training/heterogeneous-clusters/pt.grpc.sagemaker/code/requirements.txt:1]: Would like all prerequisites mentioned in main docs in the root of this project.

File: training/heterogeneous-clusters/pt.grpc.sagemaker/hetero-pytorch-mnist.ipynb [hunk 16]
```diff
@@ -0,0 +1,509 @@
...
+   "source": [
+    "## PyTorch's example to demonstrate Amazon SageMaker Heterogeneous Cluster for model training\n",
+    "\n",
+    "---\n",
+    "### Description\n",
+    "Heterogeneous clusters enable launching training jobs that use multiple instance types in a single job. This capability can improve your training cost and speed by running different parts of the model training on the most suitable instance type. This use case typically happens in computer vision DL training, where training is bottleneck on CPU resources needed for data augmentation, leaving the expensive GPU underutilized. Heterogeneous clusters enable you to add more CPU resources to fully utilize GPUs, thus increase training speed and cost-efficiency. For more details, you can find the documentation of this feature [here](https://docs.aws.amazon.com/sagemaker/latest/dg/train-heterogeneous-cluster.html).\n",
+    "\n",
+    "This notebook demonstrates how to use Heterogeneous Cluster feature of SageMaker Training with PyTorch 1.10.  The notebook works on Python 3 (_PyTorch 1.10 Python 3.8 CPU Optimized_) image of SageMaker Studio Notebook instance, and runs on _ml.t3.medium_ instance type.\n",
```
Comment (aaronmarkham) [inline training/heterogeneous-clusters/pt.grpc.sagemaker/hetero-pytorch-mnist.ipynb:14]: You're using 1.11 everywhere else. Why is this 1.10?

File: training/heterogeneous-clusters/pt.grpc.sagemaker/hetero-pytorch-mnist.ipynb [hunk 17]
```diff
@@ -0,0 +1,509 @@
...
+    "\n",
+    "In this sample notebook, we have taken the PyTorch model based on this [official MNIST example](https://github.com/pytorch/examples/tree/main/MNIST). We modified the training code to be heavy on data pre-processing. We are going to train this model in both Homogeneous and Heterogeneous Cluster modes. The flag to train on any of these modes can be set using `IS_HETERO = False or True` in section **B.2 Configure environment variables**. \n",
+    "\n",
+    "<table>\n",
+    "  <tr>\n",
+    "    <td width=43%><b>Homogeneous training job</b><br/>\n",
+    "    We first run the homogeneous training job to get baseline performance number, and observe that it is unable to fully utilize GPU of ml.p3.2xlarge instance due to a CPU being the bottleneck.</td>\n",
```
Comment (aaronmarkham) [inline training/heterogeneous-clusters/pt.grpc.sagemaker/hetero-pytorch-mnist.ipynb:21]: These instance types are not always available across the major commercial regions.

File: training/heterogeneous-clusters/pt.grpc.sagemaker/hetero-pytorch-mnist.ipynb [hunk 18]
```diff
@@ -0,0 +1,509 @@
...
+    "python3 -m pip install --upgrade boto3 botocore awscli sagemaker"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "0d20b2f3",
+   "metadata": {},
+   "source": [
+    "#### Step 2 - Restart the notebook kernel \n",
+    "From the Jupyter Lab menu bar **Kernel > Restart Kernel...**"
```
Comment (aaronmarkham) [inline training/heterogeneous-clusters/pt.grpc.sagemaker/hetero-pytorch-mnist.ipynb:96]: You should do this programmatically for the sake of testing.

File: training/heterogeneous-clusters/pt.grpc.sagemaker/hetero-pytorch-mnist.ipynb [hunk 19]
```diff
@@ -0,0 +1,509 @@
...
+   "cell_type": "code",
+   "execution_count": 53,
+   "id": "594fce53",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "arn:aws:iam::776941257690:role/service-role/AmazonSageMakerServiceCatalogProductsUseRole\n",
```
Comment (aaronmarkham) [inline training/heterogeneous-clusters/pt.grpc.sagemaker/hetero-pytorch-mnist.ipynb:162]: Think you should clear your outputs. If you need to keep some outputs we can look into Papermill or something to hide some things and display others, but keep functionality for testing.

File: training/heterogeneous-clusters/pt.grpc.sagemaker/hetero-pytorch-mnist.ipynb [hunk 20]
```diff
@@ -0,0 +1,509 @@
...
+   "outputs": [],
+   "source": [
+    "import logging\n",
+    "import boto3\n",
+    "from botocore.exceptions import ClientError\n",
+    "\n",
+    "# Download training and testing data from a public S3 bucket\n",
+    "\n",
+    "\n",
+    "def download_from_s3(data_dir=\"./data\", train=True):\n",
```
Comment (aaronmarkham) [inline training/heterogeneous-clusters/pt.grpc.sagemaker/hetero-pytorch-mnist.ipynb:334]: These helper functions exist in the [Python SDK](https://sagemaker.readthedocs.io/en/stable/api/utility/s3.html).

**Review by aaronmarkham — CHANGES_REQUESTED** [review]
Very extensive example! I'm concerned about testing it though. Some of the example is just in readme which isn't testable. Also, there's hardcoding of region to us-east-1 and of instance types which will fail when we test in other regions. There's a lot of security issues that the bots have picked up, but I'd also say if you're setting up network comms like this, you need a security section in the overview of the solution(s). With the complexity of the solution, I can't help but think 1) could a lot of this be done in a container to reduce the amount of noise and complexity for someone trying this out; 2) if untestable across regions, then should it be moved to aws-samples and we can simply link to it from the examples website (with caveats that it goes untested). If you can work through the security issues, and solve the problem for several/many/most commercial regions, then I think we could proceed. But locking it to us-east-1: I don't think is great customer experience.

File: training/heterogeneous-clusters/hello.world.sagemaker/start_job.py (hunk 4 above)
Comment (gilinachum) [reply training/heterogeneous-clusters/hello.world.sagemaker/start_job.py:6]: I'm removing all references to us-east-1. The notebooks should work in all regions that with the instance types we test/benchmark with: ml.g5.*, and ml.p4d.24xlarge.
Heterogeneous clusters feature specifically focuses and improve highend compute instances with fixed amount of CPU:GPU, therefore we benchmark with a P4d which is the best candidate for users of the feature, however, a user could change to other instance types as they see fit.

**Comment by gilinachum** [comment]
> Very extensive example! I'm concerned about testing it though. Some of the example is just in readme which isn't testable. Also, there's hardcoding of region to us-east-1 and of instance types which will fail when we test in other regions. There's a lot of security issues that the bots have picked up, but I'd also say if you're setting up network comms like this, you need a security section in the overview of the solution(s). With the complexity of the solution, I can't help but think 1) could a lot of this be done in a container to reduce the amount of noise and complexity for someone trying this out; 2) if untestable across regions, then should it be moved to aws-samples and we can simply link to it from the examples website (with caveats that it goes untested). If you can work through the security issues, and solve the problem for several/many/most commercial regions, then I think we could proceed. But locking it to us-east-1: I don't think is great customer experience.

@aaronmarkham 
- Testing all code - Except for the root level readme we won't have readme files (we'll link directly to notebooks). We'll also make sure all of the runnable code runs as part of these notebooks. I'd like to keep the solution in SM Examples repo since it's the best, and currently the only, representation to the Heterogeneous clusters feature of SM Training.
- Security communication channels - We'll add a section to the main readme that explains the network connections between the instances, port numbers and their purpose. Unlike inference, which is exposed (directly or indirectly) to end users, this is a training workload, where all the software is controlled by the data scientists that runs training. The solution communicates between training instances (all ports communication between instances is allowed by default), just like other processes that takes place during distributed training (MPI, DDP, Horovod, SSH, etc). We don't add IAM rules, security groups, VPC, etc. 
- Security creating processes - This CodeGuru warnings are out of context, as the added code acts as a proxy script and re-executes the data scientists commands, same commands that are executed for any training job. There's no end-user input like in webapp scenario.
Normally:
user --cmd-->  [train.py]
With launcher.py acting as a proxy: 
user --cmd--> [Launcher.py] --cmd+extra_arg--> [train.py]

**Review by aaronmarkham — APPROVED** [review]
